compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs

```diff
@@ -2441,33 +2334,108 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             trait_impls
         };
 
+        let krate = self.tcx.crate_name(trait_pred.def_id().krate);
+        let name = self.tcx.item_name(trait_pred.def_id());
         let required_trait_path = self.tcx.def_path_str(trait_pred.def_id());
         let traits_with_same_path: UnordSet<_> = self
             .tcx
             .visible_traits()
-            .filter(|trait_def_id| *trait_def_id != trait_pred.def_id())
+            .filter(|trait_def_id| {
+                trait_def_id.krate != trait_pred.def_id().krate
+                    && (self.tcx.def_path_str(trait_def_id) == required_trait_path
+                        || self.tcx.crate_name(trait_def_id.krate) == krate
+                            && self.tcx.item_name(trait_def_id) == name)
```

Comment:
when is `self.tcx.crate_name(trait_def_id.krate) == krate && self.tcx.item_name(trait_def_id) == name` false but `self.tcx.def_path_str(trait_def_id) == required_trait_path` holds?

Reply:
It used by `tests/ui/traits/bound/same-crate-name.rs` , disabling this case makes the test fail.

Reply:
I have investigated. This test checks for ambiguities regarding the full import path, the import path is the same for both crates here, while the crates name are differents. The first crate name is `crate_a1`, the second crate name is `crate_a2` the item name is `Bar. So we have `crate_a1::Bar`, `crate_a2::Bar`, in both cases the import name is `main::a::Bar`.  I could of course, update the test, but I think that what's the test is checking for might be a real use case.